### PR TITLE
Fix ZkHelixClusterVerifier related resource leakage

### PR DESCRIPTION
**Pull request by kaisun2000**
### Issues

- [x] My PR addresses the following Helix issues #1184 

### Description

- [x] Here are some details about my PR, including screenshots of any UI changes:

     ZkHelixClusterVerifier and its subclass built with ZK addres would
    create an internal ZkClient. Without properly closing the ZkClient
    or using global ZkClient in the test base. There would be four
    threads and one ZooKeeper client object leaked. The leaked threads
    would be in thousands and thus be one of the main contributor
    preventing the unit test from finishing. This pull request address
    this issue.

### Tests

- [ ] The following tests are written for this issue:

None

- [ ] The following is the result of the "mvn test" command on the appropriate module:

log ERROR on

[ERROR] Tests run: 1155, Failures: 4, Errors: 0, Skipped: 1, Time elapsed: 5,337.325 s <<< FAILURE! - in TestSuite
[ERROR] testResetSnapshots(org.apache.helix.controller.changedetector.TestResourceChangeDetector)  Time elapsed: 300.01 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.testng.internal.TestNGMethod.testResetSnapshots() didn't finish within the time-out 300000
	at org.apache.helix.controller.changedetector.TestResourceChangeDetector.testResetSnapshots(TestResourceChangeDetector.java:431)

[ERROR] testWorkflowRunningTime(org.apache.helix.integration.task.TestWorkflowTimeout)  Time elapsed: 10.068 s  <<< FAILURE!
org.apache.helix.HelixException: Workflow "testWorkflowRunningTime" context is empty or not in states: "[TIMED_OUT]", current state: "COMPLETED"
	at org.apache.helix.integration.task.TestWorkflowTimeout.testWorkflowRunningTime(TestWorkflowTimeout.java:66)

[ERROR] testFixedTargetTaskAndDisabledRebalanceAndNodeAdded(org.apache.helix.integration.task.TestRebalanceRunningTask)  Time elapsed: 10.25 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.integration.task.TestRebalanceRunningTask.testFixedTargetTaskAndDisabledRebalanceAndNodeAdded(TestRebalanceRunningTask.java:297)

[ERROR] testLostZkConnection(org.apache.helix.integration.TestZkConnectionLost)  Time elapsed: 300.015 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.testng.internal.TestNGMethod.testLostZkConnection() didn't finish within the time-out 300000

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestResourceChangeDetector.testResetSnapshots:431 » ThreadTimeout Method org.t...
[ERROR]   TestZkConnectionLost.testLostZkConnection » ThreadTimeout Method org.testng.in...
[ERROR]   TestRebalanceRunningTask.testFixedTargetTaskAndDisabledRebalanceAndNodeAdded:297 expected:<true> but was:<false>
[ERROR]   TestWorkflowTimeout.testWorkflowRunningTime:66 » Helix Workflow "testWorkflowR...
[INFO] 
[ERROR] Tests run: 1155, Failures: 4, Errors: 0, Skipped: 1
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:29 h
[INFO] Finished at: 2020-07-30T22:21:18-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/ksun/dev_branch_helix/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
ksun-mn1:helix-core ksun$ 

---------------
log INFO on

[ERROR] Tests run: 1154, Failures: 16, Errors: 0, Skipped: 8, Time elapsed: 6,344.539 s <<< FAILURE! - in TestSuite
[ERROR] testDisableDelayRebalanceInCluster(org.apache.helix.integration.rebalancer.WagedRebalancer.TestDelayedWagedRebalanceWithDisabledInstance)  Time elapsed: 1.26 s  <<< FAILURE!
java.lang.AssertionError: Test-DB-createTestDBs0's best possible assignment is null for partition Test-DB-createTestDBs0_0 expected:<true> but was:<false>
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestDelayedWagedRebalanceWithDisabledInstance.testDisableDelayRebalanceInCluster(TestDelayedWagedRebalanceWithDisabledInstance.java:88)

[ERROR] testPartitionMovementAfterDelayTime(org.apache.helix.integration.rebalancer.WagedRebalancer.TestDelayedWagedRebalanceWithDisabledInstance)  Time elapsed: 1.076 s  <<< FAILURE!
java.lang.AssertionError: Test-DB-createTestDBs0's best possible assignment is null for partition Test-DB-createTestDBs0_0 expected:<true> but was:<false>
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestDelayedWagedRebalanceWithDisabledInstance.testPartitionMovementAfterDelayTime(TestDelayedWagedRebalanceWithDisabledInstance.java:82)

[ERROR] testAddZone(org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalanceFaultZone)  Time elapsed: 0.54 s  <<< FAILURE!
java.lang.NullPointerException
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalanceFaultZone.validateZoneAndTagIsolation(TestWagedRebalanceFaultZone.java:327)
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalanceFaultZone.validate(TestWagedRebalanceFaultZone.java:315)
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalanceFaultZone.testAddZone(TestWagedRebalanceFaultZone.java:239)

[ERROR] testRebalancerReset(org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalance)  Time elapsed: 5.844 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalance.validate(TestWagedRebalance.java:646)
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalance.testRebalancerReset(TestWagedRebalance.java:611)

[ERROR] testAddZone(org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalanceTopologyAware)  Time elapsed: 0.633 s  <<< FAILURE!
java.lang.NullPointerException
	at org.apache.helix.integration.rebalancer.WagedRebalancer.TestWagedRebalanceTopologyAware.testAddZone(TestWagedRebalanceTopologyAware.java:112)

[ERROR] testClusterExpansion(org.apache.helix.integration.rebalancer.PartitionMigration.TestExpandCluster)  Time elapsed: 12.021 s  <<< FAILURE!
java.lang.AssertionError: expected:<false> but was:<true>
	at org.apache.helix.integration.rebalancer.PartitionMigration.TestExpandCluster.testClusterExpansion(TestExpandCluster.java:61)

[ERROR] testLackEnoughRacks(org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalance)  Time elapsed: 1.201 s  <<< FAILURE!
java.lang.AssertionError: expected:<2> but was:<1>
	at org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalance.validateZoneAndTagIsolation(TestCrushAutoRebalance.java:293)
	at org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalance.testLackEnoughRacks(TestCrushAutoRebalance.java:259)

[ERROR] test(org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalanceTopoplogyAwareDisabled)  Time elapsed: 6.005 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalanceTopoplogyAwareDisabled.test(TestCrushAutoRebalanceTopoplogyAwareDisabled.java:66)

[ERROR] testP2PStateTransitionEnabled(org.apache.helix.integration.messaging.TestP2PNoDuplicatedMessage)  Time elapsed: 39.132 s  <<< FAILURE!
java.lang.AssertionError: expected:<2200> but was:<1754>
	at org.apache.helix.integration.messaging.TestP2PNoDuplicatedMessage.testP2PStateTransitionEnabled(TestP2PNoDuplicatedMessage.java:174)

[ERROR] afterMethod(org.apache.helix.tools.TestClusterStateVerifier)  Time elapsed: 1.262 s  <<< FAILURE!
java.lang.IllegalStateException: ZkClient already closed!
	at org.apache.helix.tools.TestClusterStateVerifier.afterMethod(TestClusterStateVerifier.java:98)

[ERROR] beforeClass(org.apache.helix.integration.rebalancer.TestAutoRebalancePartitionLimit)  Time elapsed: 1.726 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.integration.rebalancer.TestAutoRebalancePartitionLimit.beforeClass(TestAutoRebalancePartitionLimit.java:80)

[ERROR] testEnableCompressionResource(org.apache.helix.integration.TestEnableCompression)  Time elapsed: 139.954 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.integration.TestEnableCompression.testEnableCompressionResource(TestEnableCompression.java:117)

[ERROR] testWorkflowRunningTime(org.apache.helix.integration.task.TestWorkflowTimeout)  Time elapsed: 10.073 s  <<< FAILURE!
org.apache.helix.HelixException: Workflow "testWorkflowRunningTime" context is empty or not in states: "[TIMED_OUT]", current state: "COMPLETED"
	at org.apache.helix.integration.task.TestWorkflowTimeout.testWorkflowRunningTime(TestWorkflowTimeout.java:66)

[ERROR] testLostZkConnection(org.apache.helix.integration.TestZkConnectionLost)  Time elapsed: 300.019 s  <<< FAILURE!
org.testng.internal.thread.ThreadTimeoutException: Method org.testng.internal.TestNGMethod.testLostZkConnection() didn't finish within the time-out 300000

[ERROR] test(org.apache.helix.integration.TestDisableCustomCodeRunner)  Time elapsed: 1.845 s  <<< FAILURE!
java.lang.AssertionError: expected:<false> but was:<true>
	at org.apache.helix.integration.TestDisableCustomCodeRunner.test(TestDisableCustomCodeRunner.java:233)

[ERROR] testClusterStatusMonitorLifecycle(org.apache.helix.monitoring.TestClusterStatusMonitorLifecycle)  Time elapsed: 6.693 s  <<< FAILURE!
java.lang.AssertionError: expected:<true> but was:<false>
	at org.apache.helix.monitoring.TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle(TestClusterStatusMonitorLifecycle.java:274)

[INFO] 
[INFO] Results:
[INFO] 
[ERROR] Failures: 
[ERROR]   TestDisableCustomCodeRunner.test:233 expected:<false> but was:<true>
[ERROR]   TestEnableCompression.testEnableCompressionResource:117 expected:<true> but was:<false>
[ERROR]   TestZkConnectionLost.testLostZkConnection » ThreadTimeout Method org.testng.in...
[ERROR]   TestP2PNoDuplicatedMessage.testP2PStateTransitionEnabled:174 expected:<2200> but was:<1754>
[ERROR]   TestCrushAutoRebalance.testLackEnoughRacks:259->validateZoneAndTagIsolation:293 expected:<2> but was:<1>
[ERROR] org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalanceTopoplogyAwareDisabled.test(org.apache.helix.integration.rebalancer.CrushRebalancers.TestCrushAutoRebalanceTopoplogyAwareDisabled)
[INFO]   Run 1: PASS
[ERROR]   Run 2: TestCrushAutoRebalanceTopoplogyAwareDisabled.test:66->TestCrushAutoRebalanceNonRack.test:159 expected:<true> but was:<false>
[INFO] 
[ERROR]   TestExpandCluster.testClusterExpansion:61 expected:<false> but was:<true>
[ERROR]   TestAutoRebalancePartitionLimit.beforeClass:80 expected:<true> but was:<false>
[ERROR]   TestDelayedWagedRebalanceWithDisabledInstance.testDisableDelayRebalanceInCluster:88->TestDelayedAutoRebalanceWithDisabledInstance.testDisableDelayRebalanceInCluster:270->ZkTestBase.validateMinActiveAndTopStateReplica:499 Test-DB-createTestDBs0's best possible assignment is null for partition Test-DB-createTestDBs0_0 expected:<true> but was:<false>
[ERROR]   TestDelayedWagedRebalanceWithDisabledInstance.testPartitionMovementAfterDelayTime:82->TestDelayedAutoRebalanceWithDisabledInstance.testPartitionMovementAfterDelayTime:192->ZkTestBase.validateMinActiveAndTopStateReplica:499 Test-DB-createTestDBs0's best possible assignment is null for partition Test-DB-createTestDBs0_0 expected:<true> but was:<false>
[ERROR]   TestWagedRebalance.testRebalancerReset:611->validate:646 expected:<true> but was:<false>
[ERROR]   TestWagedRebalanceFaultZone.testAddZone:239->validate:315->validateZoneAndTagIsolation:327 NullPointer
[ERROR]   TestWagedRebalanceTopologyAware.testAddZone:112->TestWagedRebalanceFaultZone.testAddZone:239->TestWagedRebalanceFaultZone.validate:315->TestWagedRebalanceFaultZone.validateZoneAndTagIsolation:328 » NullPointer
[ERROR]   TestWorkflowTimeout.testWorkflowRunningTime:66 » Helix Workflow "testWorkflowR...
[ERROR]   TestClusterStatusMonitorLifecycle.testClusterStatusMonitorLifecycle:274 expected:<true> but was:<false>
[ERROR]   TestClusterStateVerifier.afterMethod:98 » IllegalState ZkClient already closed...
[INFO] 
[ERROR] Tests run: 1153, Failures: 16, Errors: 0, Skipped: 8
[INFO] 
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  01:45 h
[INFO] Finished at: 2020-07-30T16:47:39-07:00
[INFO] ------------------------------------------------------------------------
[ERROR] Failed to execute goal org.apache.maven.plugins:maven-surefire-plugin:3.0.0-M3:test (default-test) on project helix-core: There are test failures.
[ERROR] 
[ERROR] Please refer to /Users/ksun/dev_branch_helix/helix/helix-core/target/surefire-reports for the individual test results.
[ERROR] Please refer to dump files (if any exist) [date].dump, [date]-jvmRun[N].dump and [date].dumpstream.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoFailureException
ksun-mn1:helix-core ksun$ mvn test 

### Commits

- [ ] My commits all reference appropriate Apache Helix GitHub issues in their subject lines. In addition, my commits follow the guidelines from "[How to write a good git commit message](http://chris.beams.io/posts/git-commit/)":
  1. Subject is separated from body by a blank line
  1. Subject is limited to 50 characters (not including Jira issue reference)
  1. Subject does not end with a period
  1. Subject uses the imperative mood ("add", not "adding")
  1. Body wraps at 72 characters
  1. Body explains "what" and "why", not "how"

### Documentation (Optional)

- [ ] In case of new functionality, my PR adds documentation in the following wiki page:

(Link the GitHub wiki you added)

### Code Quality

- [ ] My diff has been formatted using helix-style.xml 
(helix-style-intellij.xml if IntelliJ IDE is used)

### CODE REVIEW

File: helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java [hunk 1]
```diff
@@ -74,7 +81,8 @@ public void syncStart() {
 
     _watcher = new Thread(this);
     _watcher.setName(String
-        .format("ClusterManager_Watcher_%s_%s_%s", _clusterName, _instanceName, _type.name()));
+        .format("ClusterManager_Watcher_%s_%s_%s_%d", _clusterName, _instanceName, _type.name(), _uid));
+    LOG.info("ClusterManager_watcher_{}_{}_{}_{} started, stacktrace {}", _clusterName, _instanceName, _type.name(), _uid, Thread.currentThread().getStackTrace());
```
Comment (pkuwm) [inline helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:85]: It seems this log is for your debugging. Do we want to remove it then?

Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:85]: No, we don't. We specifically need this.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java [hunk 2]
```diff
@@ -120,6 +122,41 @@ public void handle() {
     }
   }
 
+  public static class SyncCallbackHandler extends DefaultCallback implements AsyncCallback.VoidCallback {
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/callback/ZkAsyncCallbacks.java]: rebased.

File: helix-lock/src/test/java/org/apache/helix/lock/helix/TestZKHelixNonblockingLock.java [hunk 3]
```diff
@@ -76,6 +76,7 @@ public void beforeMethod() {
 
   @AfterSuite
   public void afterSuite() throws IOException {
+    System.out.println("afterSuite " + getClass().getName());
```
Comment (jiajunwang) [inline helix-lock/src/test/java/org/apache/helix/lock/helix/TestZKHelixNonblockingLock.java:79]: Same here, what's the usage of this output?

File: helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java (hunk 1 above)
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:85]: This is testing code. Make it info won't pollute log. Will change to debug anyway.

File: helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java [hunk 4]
```diff
@@ -200,6 +200,7 @@ private void startZooKeeper(int i)
 
   @AfterSuite
   public void afterSuite() throws IOException {
+    System.out.println("afterSuite " + getClass().getName());
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/common/ZkTestBase.java:203]: Removed.

File: helix-core/src/test/java/org/apache/helix/integration/TestDriver.java [hunk 5]
```diff
@@ -246,7 +246,11 @@ public static void verifyCluster(String uniqClusterName, long beginTime, long ti
 
     ZkHelixClusterVerifier verifier =
         new BestPossibleExternalViewVerifier.Builder(clusterName).setZkAddr(ZK_ADDR).build();
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/TestDriver.java:248]: Note, here, in TestDriver, we don't have _gZkClient as it is not a subclass of ZkTestBase.java

File: helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java [hunk 6]
```diff
@@ -51,6 +55,9 @@ protected ClusterManager(String zkAddr, String clusterName, String instanceName,
     _clusterName = clusterName;
     _instanceName = instanceName;
     _type = type;
+    _uid = UID.getAndIncrement();
+
+    LOG.info("ClusterManager_watcher_{}_{}_{}_{} created, stacktrace {}", _clusterName, _instanceName, _type.name(), _uid, Thread.currentThread().getStackTrace());
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/test/java/org/apache/helix/integration/manager/ClusterManager.java:60]: You are right. Originally, the idea is that sync_start time, you don't see which one is created as you want the creation context of the object, not the creation context of the starting thread. Later, it turned the two places are the same test class. so it does not really matter. Let me remove this one.

File: helix-core/src/test/resources/log4j.properties [hunk 7]
```diff
@@ -17,7 +17,7 @@
 # under the License.
 #
 # Set root logger level to DEBUG and its only appender to R.
-log4j.rootLogger=ERROR, C
+log4j.rootLogger=ERROR, C, R
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/test/resources/log4j.properties:20]: That is the idea. Even for current test, in order to record the log, other wise, you need to do something like ` mvn test > test.log 2>&1 ` without this one.

[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/test/resources/log4j.properties:20]: anyway, let me remove it for now. Let us get this diff in to trunk first.

File: helix-core/src/test/resources/log4j.properties [hunk 8]
```diff
@@ -37,5 +38,6 @@ log4j.logger.org.I0Itec=ERROR
 log4j.logger.org.apache=ERROR
 log4j.logger.com.noelios=ERROR
 log4j.logger.org.restlet=ERROR
+#log4j.logger.org.apache.helix=INFO
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply helix-core/src/test/resources/log4j.properties:41]: removed.

File: helix-lock/src/test/java/org/apache/helix/lock/helix/TestZKHelixNonblockingLock.java (hunk 3 above)
Comment (kaisun2000) [reply helix-lock/src/test/java/org/apache/helix/lock/helix/TestZKHelixNonblockingLock.java:79]: removed.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 9]
```diff
@@ -215,13 +218,21 @@ protected ZkClient(IZkConnection zkConnection, int connectionTimeout, long opera
       throw new NullPointerException("Zookeeper connection is null!");
     }
 
+    _uid = UID.getAndIncrement();
+    if (LOG.isInfoEnabled()) {
+      LOG.info("ZkClient created with _uid {}, stacktrace {}", _uid, Thread.currentThread().getStackTrace());
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:223]: There are many time when I debugged, I wish to know where the zkclient is created when I look at the zkclient event thread logs. I think this is actually very useful. In fact, I would event want to add the _uid to each line of logs from Zkclient. It would save a lot time.

[the remark this answers is not in the document]
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:223]: Basically, for a log from a thread, such as ZkClient event thread or asycn retry thread, you have the thread id in the log. grepping thread id, you will find the _uid, grepping the _uid, you see the context of where the zkclient is created, such from zkhelixmanager or whatever other Helix java api. This is basically also the technique used to track the watcher leakage. 

The log would not be excessive as only one zkclient creation has 3 more lines.

For now, let me change it to debug, if this is ok. We can discussed adding this thing to callbackhandler and zkclient later.

[the remark this answers is not in the document]
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:223]: removed.

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 10]
```diff
@@ -215,13 +218,21 @@ protected ZkClient(IZkConnection zkConnection, int connectionTimeout, long opera
...
+
     _connection = zkConnection;
     _pathBasedZkSerializer = zkSerializer;
     _operationRetryTimeoutInMillis = operationRetryTimeout;
     _isNewSessionEventFired = false;
 
     _asyncCallRetryThread = new ZkAsyncRetryThread(zkConnection.getServers());
     _asyncCallRetryThread.start();
+    if (LOG.isInfoEnabled()) {
+      LOG.info("ZkClient created with _uid {}, _asyncCallRetryThread id {}", _uid, _asyncCallRetryThread.getId());
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:234]: changed to debug

File: zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java [hunk 11]
```diff
@@ -2157,6 +2166,8 @@ public void connect(final long maxMsToWaitUntilConnected, Watcher watcher)
       _eventThread = new ZkEventThread(zkConnection.getServers());
       _eventThread.start();
 
+      LOG.info("ZkClient created with _uid {}, _eventThread {}", _uid, _eventThread.getId());
```
[the remark this answers is not in the document]
Comment (kaisun2000) [reply zookeeper-api/src/main/java/org/apache/helix/zookeeper/zkclient/ZkClient.java:2169]: changed to debug

File: helix-core/src/test/java/org/apache/helix/integration/common/ZkStandAloneCMTestBase.java [hunk 12]
```diff
@@ -81,7 +81,8 @@ public void beforeClass() throws Exception {
     _controller = new ClusterControllerManager(ZK_ADDR, CLUSTER_NAME, controllerName);
     _controller.syncStart();
 
-    _clusterVerifier = new BestPossibleExternalViewVerifier.Builder(CLUSTER_NAME).setZkAddr(ZK_ADDR).build();
+    //_clusterVerifier = new BestPossibleExternalViewVerifier.Builder(CLUSTER_NAME).setZkAddr(ZK_ADDR).build();
```
Comment (jiajunwang) [inline helix-core/src/test/java/org/apache/helix/integration/common/ZkStandAloneCMTestBase.java:84]: Remove?

File: helix-core/src/test/resources/log4j.properties [hunk 13]
```diff
@@ -27,6 +27,7 @@ log4j.appender.R=org.apache.log4j.RollingFileAppender
 log4j.appender.R.layout=org.apache.log4j.PatternLayout
 log4j.appender.R.layout.ConversionPattern=%5p [%C:%M] (%F:%L) - %m%n
 log4j.appender.R.File=target/ClusterManagerLogs/log.txt
+log4j.appender.R.MaxBackupIndex=20
```
Comment (jiajunwang) [inline helix-core/src/test/resources/log4j.properties:30]: What's the motivation for this change? Did we hit the limitation?
Once we have the test log cleaned up, we will have a smaller log and fit the default file size, right?

Comment (kaisun2000) [reply helix-core/src/test/resources/log4j.properties:30]: removed.

**Comment by kaisun2000** [comment]
> What is helix-core/out? Test result?
> This is why I suggest we don't configure the test log to output to any file.

Good catch. Too many file changed, did not realize I also accidentally added the test log file

**Comment by kaisun2000** [comment]
This pull request is approved, please help to merge.

>ZkHelixClusterVerifier and its subclass built with ZK addres would
create an internal ZkClient. Without properly closing the ZkClient
or using global ZkClient in the test base. There would be four
threads and one ZooKeeper client object leaked. The leaked threads
would be in thousands and thus be one of the main contributor
preventing the unit test from finishing. This pull request address
this issue.
